models/contacts.js

```diff
@@ -1,14 +1,97 @@
-// const fs = require('fs/promises')
+const fs = require('fs/promises');
+const path = require('path');
+const Joi = require("joi");
+const { v4: uuidv4 } = require("uuid");
 
-const listContacts = async () => {}
+console.log("here is my language")
 
-const getContactById = async (contactId) => {}
+const contactsPath = path.format({ dir: "./models", base: "contacts.json"});
 
-const removeContact = async (contactId) => {}
 
-const addContact = async (body) => {}
+const listContacts = async (filePath, arr) => {
+  const string = JSON.stringify(arr, null, 2);
+  try {
+    await fs.writeFile(filePath, string);
+  } catch (err) {
+    console.error(err)
+  }
+};
 
-const updateContact = async (contactId, body) => {}
+const getContactById = async (contactId) => {
+  const contactData = await listContacts();
+  const contact = contactData.filter(({ id }) => id === contactId);
+  if (contact.length === 0) {
+    throw new Error("Contact not found");
+  }
+  return contact;
+};
+
+const removeContact = async (contactId) => {
+  const contactsData = await listContacts();
+  const contactIndex = contactsData.findIndex(
+    (contact) => contact.id === contactId
+  );
+
+  if (contactIndex === -1) {
+    throw new Error("Contact not found");
+  }
+
+  contactsData.splice(contactIndex, 1);
+  await saveArrayToFile(contactsPath, contactsData);
+};
+
+const addContact = async (name, email, phone) => {
+  const contacts = await listContacts();
+
+  const contact = {
+    id: uuidv4(),
+    name,
+    email,
+    phone,
+  };
+
+  const updatedContacts = [...contacts, contact];
+  await saveArrayToFile(contactsPath, updatedContacts);
+
+  return contact;
+};
+
+
+
+const updateContact = async (contactId, body) => {
+  const {name, email, phone} = body;
+
+  const contactsData = await listContacts();
+  const contactIndex = contactsData.findIndex(
+    (contact) => contact.id === contactId
+  );
+
+  if (contactIndex === -1) {
+    throw new Error("Contact not found");
+  }
+  const schema = Joi.object({
```

Comment:
Walidacja musi być zrobiona na samym początku uruchamiania funkcji. Jeśli cokolwiek będzie nie tak to zatrzymujesz wykonanie i zwracasz błąd. Jakakolwiek operacja na bazie danych ma się dziać dopiero wtedy kiedy wszystko będzie poprawne.